File: algorithms/javaalgo/src/test/java/programmers/level2/nlcm/NLCMTest.java
package programmers.level2.nlcm;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertAll;

import java.util.Arrays;
import org.junit.jupiter.api.Test;

public class NLCMTest {

    @Test
    void test() {
        int[] arr1 = {2, 6, 8, 14};
        int[] arr2 = {1, 2, 3};

        assertAll(
            () -> assertThat(solution(arr1)).isEqualTo(168),
            () -> assertThat(solution(arr2)).isEqualTo(6)
        );
    }

    public int solution(int[] arr) {
        if (arr.length == 1) {
            return arr[0];
        }
        Arrays.sort(arr);

        int gcd = gcd(arr[0], arr[1]);
        int lcm = (arr[0] * arr[1]) / gcd;

        for (int i = 2; i < arr.length; i++) {
            gcd = gcd(lcm, arr[i]);
            lcm = (lcm * arr[i]) / gcd;
        }
        return lcm;
    }

    public int gcd(int num1, int num2) {
        if (num1 % num2 == 0) {
            return num2;
        }
        return gcd(num2, num1 % num2);
    }
}

File: algorithms/javascriptalgo/programmers/level2/nLCM.test.js
test("test", () => {
    const arr1 = [2, 6, 8, 14];
    const arr2 = [1, 2, 3];

    expect(solution(arr1)).toBe(168);
    expect(solution(arr2)).toBe(6);
    expect(solution2(arr1)).toBe(168);
});

function solution(arr) {
    if (arr.length === 1) {
        return arr[0];
    }
    arr.sort();

    let gcd = makeGcd(arr[0], arr[1]);
    let lcm = (arr[0] * arr[1]) / gcd;

    for (let i = 2; i < arr.length; i++) {
        gcd = makeGcd(lcm, arr[i]);
        lcm = (lcm * arr[i]) / gcd;
    }
    return lcm;
}

function makeGcd(a, b) {
    if (a % b === 0) {
        return b;
    }
    return makeGcd(b, a % b);
}

function solution2(arr) {
    return arr.reduce((a, b) => a * b / gcd2(a, b));
}

function gcd2(a, b) {
    return a % b === 0 ? b : gcd2(b, a % b);
}
